monitoring/stats_dump_scheduler.cc

```diff
@@ -0,0 +1,98 @@
+//  Copyright (c) 2011-present, Facebook, Inc.  All rights reserved.
+//  This source code is licensed under both the GPLv2 (found in the
+//  COPYING file in the root directory) and Apache 2.0 License
+//  (found in the LICENSE.Apache file in the root directory).
+
+#include "monitoring/stats_dump_scheduler.h"
+
+#include <stdio.h>
+
+#include "db/db_impl/db_impl.h"
+#include "util/cast_util.h"
+
+#ifndef ROCKSDB_LITE
+namespace ROCKSDB_NAMESPACE {
+
+port::Mutex StatsDumpScheduler::mutex_;
+
+StatsDumpScheduler::StatsDumpScheduler(Env* env) {
+  timer = new Timer(env);
+  timer->Start();
+}
+
+StatsDumpScheduler::~StatsDumpScheduler() {
+  if (timer) {
+    timer->Shutdown();
+    delete timer;
+  }
+}
+
+void StatsDumpScheduler::Register(DB* db, unsigned int stats_dump_period_sec,
+                                  unsigned int stats_persist_period_sec) {
+  DBImpl* dbi = static_cast_with_check<DBImpl>(db);
+
+  if (stats_dump_period_sec > 0) {
+    timer->Add([dbi]() { dbi->DumpStats(); }, GetTaskName(dbi, "dump_st"), 0,
+               stats_dump_period_sec * 1e6);
+  }
+  if (stats_persist_period_sec > 0) {
+    timer->Add([dbi]() { dbi->PersistStats(); }, GetTaskName(dbi, "pst_st"), 0,
+               stats_persist_period_sec * 1e6);
+  }
+}
+
+void StatsDumpScheduler::Unregister(DB* db) {
+  DBImpl* dbi = static_cast_with_check<DBImpl>(db);
+  timer->Cancel(GetTaskName(dbi, "dump_st"));
+  timer->Cancel(GetTaskName(dbi, "pst_st"));
```

Comment:
The only thing that needs to be extremely careful is DB shut down case. We need to make sure that we only start to destroy any class after all scheduled functions are cancelled and and ongoing function finishes running. It's not clear Timer::Cancel() is sufficient here. Maybe I got it wrong.

Reply:
+1.
`timer->Cancel()` only sets a boolean flag. What we need to do here is more than that: if the function has already started running, then we need to wait for the function to finish. Otherwise, the ongoing function that will access `db` will segfault.

Reply:
As discussed offline, currently there's no race condition issue as the timer hold the mutex while running function, the issue will be fixed in https://github.com/facebook/rocksdb/pull/7228 which will also make sure it wait until the current function is finished.